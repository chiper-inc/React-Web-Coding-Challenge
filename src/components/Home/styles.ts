import { styled } from '@mui/material';

export const MainHome = styled('main')(({ theme }) => ({
  color: theme.palette.primary.contrastText,
  backgroundColor: theme.palette.primary.light,
  padding: '20px',
  height: '100%',
  boxSizing: 'border-box'
}));

export const BodyByke = styled('main')`
  margin: 0 auto;
  width: 100%;
  max-width: 1600px;
  height: calc(100% - 230px);

  form {
    width: 100%;
  }
`;
