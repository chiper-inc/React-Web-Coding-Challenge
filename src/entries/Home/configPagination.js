export const PAGE_LEN_UI = 10
export const PAGE_LEN_API = 100
