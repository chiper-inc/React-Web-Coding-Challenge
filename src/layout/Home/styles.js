import styled from 'styled-components'

export const Wrapper = styled.div`
  min-height: 100vh;
  display: grid;
  grid-template-rows: 1fr auto;
  --featured-img: linear-gradient(180deg, #fff, #262626);
  background-color: #262626;
  background-repeat: no-repeat;
  background-image: radial-gradient(50% 50% at top center, rgba(0, 0, 0, .66), #262626), var(--featured-img);
`;

export const Main = styled.main`
  display: grid;
  grid-auto-flow: column;
  grid-gap: 6px;
  grid-template-columns: 1fr 1fr;
  @media screen and (max-width: 767px) {
    grid-auto-flow: row;
    grid-template-columns: none;
  }
`;
