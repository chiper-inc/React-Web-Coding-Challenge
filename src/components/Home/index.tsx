import React, { useEffect, useState } from 'react';
import { Header } from '../Header';
import { MainHome, BodyByke } from './styles';
import { Filters } from '../Filters';
import { BykeList } from '../BykeList';
import { ItemByke } from '../ItemByke';
import { PaginationList } from "../PaginationList";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../../redux/reducers";
import { fetchBykes } from "../../redux/actions";
import useFetch from "../../hooks/useFetch";
import { SkeletoList } from "../SkeletonList";

export interface FormFilter {
    query?: string;
    readonly per_page?: number;
    page?: any,
    colors?: string,
    readonly stolenness?: string
}

const initialValueFilter: FormFilter = {
  per_page: 10,
  query: '',
  page: 1,
  colors: ''
};

function Home () {
  const dispatch = useDispatch();
  const [initialValue, setInitialValue] = useState<FormFilter>(initialValueFilter);
  // get list colors
  // @ts-ignore
  const { data: { colors } } = useFetch(`${process.env.REACT_APP_API_URL}selections/colors`);
  const { data, total, pageMax, loading, error } = useSelector((state: RootState) => state.bykes);

  useEffect(() => {
    dispatch(fetchBykes(initialValue));
  }, [initialValue]);
  /**
     * @description this function is execute when submit from Form
     */
  const onFindBykes = (values: FormFilter) => {
    setInitialValue(
      {
        ...values,
        page: 1
      }
    );
  };
    /**
     * @description this function is execute when change pagination
     * @param page
     */
  const onChangePage = (page: number) => {
    setInitialValue(
      {
        ...initialValue,
        page
      }
    );
  };

  return (
        <MainHome>
            <Header/>
            <BodyByke>
                <Filters onFindBykes={onFindBykes} colors={colors} initialValue={initialValueFilter}/>
                <p>Total Bykes: {total}</p>

                {error &&
                    <p>An error has occurred</p>
                }
                {loading &&
                    <SkeletoList/>
                }

                {!loading && !error && <>
                    <BykeList
                        bykes={data}
                        render={(item, index) => <ItemByke key={`item-id${index}`} {...item} />}
                    />
                    <PaginationList page={initialValue.page} onChangePage={onChangePage} count={pageMax || 0}/>
                </>
                }

            </BodyByke>
        </MainHome>
  );
}

export { Home };
