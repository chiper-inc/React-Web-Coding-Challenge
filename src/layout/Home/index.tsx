import { ReactNode } from "react";
import { Footer } from "../../containers/Footer";
import { Wrapper, Main } from "./styles";

interface IProps {
  children: ReactNode
};

const HomeLayout = ({ children }: IProps) => {
  return (
    <Wrapper>
      <Main>
        {children}
      </Main>
      <Footer />
    </Wrapper>
  );
};

export default HomeLayout;
