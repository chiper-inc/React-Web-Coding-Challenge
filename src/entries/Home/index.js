import React, { useEffect } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { StolenBikesContainer } from "../../components/StolenBikesContainer";
import { requestStolenBikes, requestPage } from "../../store/actionCreators";
import { PAGE_LEN_API, PAGE_LEN_UI } from "./configPagination";

const getTotal = (bikes) =>  bikes.length >= PAGE_LEN_API ? `+${PAGE_LEN_API}`: bikes.length;

const StolenBikesContainerC = ({
  loading,
  error,
  stolenBikes,
  page,
  actions,
}) => {
  useEffect(() => {
    actions.requestStolenBikes(1,PAGE_LEN_API);
  }, [actions]);

  const handleChangePage = (page) => {
    console.log(page)
    actions.requestPage(page);
  }
  return (
    <StolenBikesContainer
      bikes={stolenBikes}
      page={page}
      lengthPage={PAGE_LEN_UI}
      total={getTotal(stolenBikes)}
      onChangePage={handleChangePage}
      loading={loading}
      error={error}
    />
  );
};

const mapStateToProps = (state) => {
  const { loading, error, stolenBikes, page, lengthPage } = state;

  return {
    loading,
    error,
    page,
    lengthPage,
    stolenBikes,
  };
};

const mapDispatchToProps = (dispatch) => ({
  actions: bindActionCreators({ requestStolenBikes, requestPage }, dispatch),
});

export const Home = connect(
  mapStateToProps,
  mapDispatchToProps
)(StolenBikesContainerC);
