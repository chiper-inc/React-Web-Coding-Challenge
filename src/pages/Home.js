import React, { useContext } from "react";
import { Grid, Typography } from "@material-ui/core";
import { SearchForm } from "../components/SearchForm";
import { Paginate } from "../components/Paginate";
import { Error } from "../components/Error";
import { Loading } from "../components/Loading";
import { ListOfIncidents } from "../components/ListOfIncidents";
import { usePaginate } from "../hooks/usePaginate";
import IncidentsContext from "../context/IncidentsContext";
import { NoData } from "../components/NoData";

export const Home = () => {
  const { page, handleChangePage } = usePaginate();

  const { loading, error, incidents } = useContext(IncidentsContext);

  return (
    <>
      <Grid container spacing={2}>
        <Grid item xs={12}>
          <SearchForm page={page} />
        </Grid>
      </Grid>
      <Grid container spacing={2}>
        <Grid item xs={12}>
          <Typography>Total 180</Typography>
        </Grid>
      </Grid>
      <Grid container spacing={2}>
        <Grid item xs={12}>
          {error !== null ? (
            <Error />
          ) : loading ? (
            <Loading />
          ) : incidents.length === 0 ? (
            <NoData />
          ) : (
            <ListOfIncidents incidents={incidents} />
          )}
        </Grid>
      </Grid>
      <Grid container spacing={2}>
        <Grid item xs={12}>
          {!loading && <Paginate currentPage={page} count={18} handleChangePage={handleChangePage} />}
        </Grid>
      </Grid>
    </>
  );
};
