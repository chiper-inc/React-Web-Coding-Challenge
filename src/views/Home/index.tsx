import HomeLayout from "../../layout/Home";
import { Title } from "../../components/Title";
import { BtnLink } from "../../components/BtnLink";
import bike from "../../assets/images/bike.svg";
import { Img, Section } from "./styles";

const HomeView = () => {
  return (
    <HomeLayout>
      <>
        <Section>
          <Img src={bike} />
        </Section>
        <Section>
          <Title text="Stolen Bike register" />
          <BtnLink to="/cases" text="Go to reports" />
        </Section>
      </>
    </HomeLayout>
  );
};

export default HomeView;
