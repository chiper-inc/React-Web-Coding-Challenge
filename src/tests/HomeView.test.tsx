import React from 'react';
import { shallow } from 'enzyme';
import Home from '../views/home';
test('renders the component Home View', () => {
  const component = shallow(<Home />);
  expect(component).toMatchSnapshot();
});
