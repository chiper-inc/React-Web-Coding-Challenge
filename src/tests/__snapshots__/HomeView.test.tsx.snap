// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component Home View 1`] = `
<div
  className="home"
>
  <BarFilter
    setTitleSearch={[Function]}
    titleSearch=""
  />
  <Loading />
</div>
`;
